stdlib/public/core/CString.swift

```diff
@@ -42,22 +42,84 @@ extension String {
   ///     }
   ///     // Prints "Caf�"
   ///
-  /// - Parameter cString: A pointer to a null-terminated UTF-8 code sequence.
-  public init(cString: UnsafePointer<CChar>) {
-    let len = UTF8._nullCodeUnitOffset(in: cString)
+  /// - Parameter nullTerminatedUTF8: A pointer to a null-terminated UTF-8 code sequence.
+  public init(cString nullTerminatedUTF8: UnsafePointer<CChar>) {
+    let len = UTF8._nullCodeUnitOffset(in: nullTerminatedUTF8)
     self = String._fromUTF8Repairing(
-      UnsafeBufferPointer(start: cString._asUInt8, count: len)).0
+      UnsafeBufferPointer(start: nullTerminatedUTF8._asUInt8, count: len)).0
+  }
+
+  @inlinable
+  @_alwaysEmitIntoClient
+  public init(cString nullTerminatedUTF8: [CChar]) {
+    self = nullTerminatedUTF8.withUnsafeBytes {
+      String(_checkingCString: $0.assumingMemoryBound(to: UInt8.self))
```

Comment:
> String(_checkingCString: $0.assumingMemoryBound(to: UInt8.self))

This looks like undefined behavior. The optimizer is still super conservative with pointer types, so it's not causing miscompiles yet.

It's sad that _fromUTF8Repairing doesn't just take a raw pointer. But if you must view `CChar` memory as `UInt8`, then `withMemoryRebound` is perfect. 

Maybe as a separate PR, I implore you to remove `_asUInt8`, which never works.

Reply:
Agreed. Removing `_asUInt8` and `_asCChar` would be a good thing, but there are enough uses scattered throughout that it should be its own PR. (rdar://91343475)